webrender_api/src/api.rs

```diff
@@ -1269,34 +1268,41 @@ pub trait NotificationHandler : Send + Sync {
     fn notify(&self, when: Checkpoint);
 }
 
-#[derive(Clone)]
 pub struct NotificationRequest {
-    handler: Arc<NotificationHandler>,
+    handler: Option<Box<NotificationHandler>>,
     when: Checkpoint,
-    done: bool,
 }
 
 impl NotificationRequest {
-    pub fn new(when: Checkpoint, handler: Arc<NotificationHandler>) -> Self {
+    pub fn new(when: Checkpoint, handler: Box<NotificationHandler>) -> Self {
         NotificationRequest {
-            handler,
+            handler: Some(handler),
             when,
-            done: false,
         }
     }
 
     pub fn when(&self) -> Checkpoint { self.when }
 
     pub fn notify(mut self) {
-        self.handler.notify(self.when);
-        self.done = true;
+        if let Some(handler) = self.handler.take() {
+            handler.notify(self.when);
+        }
     }
 }
 
 impl Drop for NotificationRequest {
     fn drop(&mut self) {
-        if !self.done {
-            self.handler.notify(Checkpoint::TransactionDropped);
+        if let Some(ref mut handler) = self.handler {
+            handler.notify(Checkpoint::TransactionDropped);
+        }
+    }
+}
+
+impl Clone for NotificationRequest {
```

Comment:
let's have some comments on why this is needed and why it's safe to clone handler as `None`. Maybe we can have a more specific method instead of using the generic `clone()`?

Reply:
I added a comment. I'm neutral about whether or not we should have a more specific method (ideally we wouldn't have to clone the messages at all). If we don't use `Clone` we'll have to manually implement the new thing for a lot of types which can be tedious or maybe the custom derive mechanism (which I don't know much about) can help. Anyway, I'd prefer that this be done separately, There's some Gecko stuff I'm working on that need this PR.